File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/java/string-algorithms.java
/**
 * Java Implementation: String Algorithms (KMP, Z-Algorithm, Aho-Corasick)
 * Invariant: Linear time O(N + M) pattern matching
 */
package seed.dsa.core;

import java.util.*;

public class StringAlgorithmsKMPZAlgorithmAhoCorasick<T> {
    private final List<T> items = new ArrayList<>();

    // TODO: Implement your solution here
    public void add(T value) {
        items.add(value);
    }

    public int size() {
        return items.size();
    }
}

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/csharp/string-algorithms.cs
/**
 * C# Implementation: String Algorithms (KMP, Z-Algorithm, Aho-Corasick)
 * Invariant: Linear time O(N + M) pattern matching
 */
using System;
using System.Collections.Generic;

namespace Seed.DSA.Core
{
    public class StringAlgorithmsKMPZAlgorithmAhoCorasick<T>
    {
        private readonly List<T> _items = new List<T>();

        // TODO: Implement algorithm operations
        public void Add(T item)
        {
            _items.Add(item);
        }

        public int Count => _items.Count;
    }
}
